code/espurna/light.cpp

```diff
@@ -607,27 +647,106 @@ unsigned int _toPWM(unsigned char id) {
     return _toPWM(_light_channels[id].current, useGamma, _light_channels[id].inverse);
 }
 
-void _lightTransition(unsigned long step) {
+namespace {
 
-    // Transitions based on current step. If step == 0, then it is the last transition
-    for (auto& channel : _light_channels) {
-        if (!step) {
-            channel.current = channel.target;
-        } else {
-            channel.current += (double) (channel.target - channel.current) / (step + 1);
+class ChannelTransition {
+public:
+    struct Transition {
+        float& value;
+        unsigned char target;
+        float step;
+        size_t count;
+
+        void debug() const {
+            DEBUG_MSG_P(PSTR("[LIGHT] Transition from=%s to=%u step=%s times=%u\n"),
+                String(value).c_str(), target, String(step).c_str(), count);
         }
+    };
+
+    explicit ChannelTransition(std::vector<channel_t>& channels, unsigned long time, unsigned long step) :
+        _time(time),
+        _step(step)
+    {
+        bool action { false };
+        for (auto& channel : channels) {
+            float diff = static_cast<float>(channel.target) - channel.current;
+            if (!time || (std::abs(diff) <= std::numeric_limits<float>::epsilon())) {
+                channel.current = channel.target;
+                continue;
+            }
+
+            action = true;
+
+            float channel_step = (diff > 0.0) ? 1.0f : -1.0f;
+            float every = static_cast<double>(time) / std::abs(diff);
+            if (every < _step) {
+                channel_step *= (static_cast<float>(_step) / every);
+                every = static_cast<float>(_step);
+            }
+            size_t count = time / every;
```

Comment:
This also changes the LIGHT_PROVIDER_DIMMER transitions, but I have not tested that yet